packages/tokenbank_v1/contracts/modules/security/RecoveryModule.sol

```diff
@@ -0,0 +1,155 @@
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "../../lib/AddressUtil.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/SignatureUtil.sol";
+
+import "../../thirdparty/ERC1271.sol";
+
+import "../../iface/Wallet.sol";
+
+import "./SecurityModule.sol";
+
+
+/// @title RecoveryModule
+contract RecoveryModule is SecurityModule
+{
+
+    event RecoveryStarted   (address indexed wallet, address indexed newOwner, uint completeAfter);
+    event RecoveryCompleted (address indexed wallet, address indexed newOwner);
+    event RecoveryCancelled (address indexed wallet);
+
+    struct WalletRecovery {
+        address newOwner;
+        uint    completeAfter;
+        uint    guardianCount;
+    }
+
+    mapping (address => WalletRecovery) public wallets;
+    uint public recoveryPeriod;
+    uint public lockPeriod;
+
+    constructor(
+        SecurityStorage _securityStorage,
+        uint _recoveryPeriod,
+        uint _lockPeriod
+        )
+        public
+        SecurityModule(_securityStorage)
+    {
+        // TODO(which should be large?)
+        recoveryPeriod = _recoveryPeriod;
+        lockPeriod = _lockPeriod;
+    }
+
+    function startRecovery(
+        address   wallet,
+        address[] calldata signers,
+        address   newOwner
+        )
+        external
+        nonReentrant
+        onlyFromMetaTx
+        notWalletOwner(wallet, newOwner)
+    {
+        require(newOwner != address(0), "ZERO_ADDRESS");
+
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter == 0, "ALREAY_STARTED");
+
+        uint guardianCount = securityStorage.numGuardians(wallet);
+        require(signers.length >= (guardianCount + 1)/2, "NOT_ENOUGH_SIGNER");
+        require(isWalletOwnerOrGuardian(wallet, signers), "UNAUTHORIZED");
+
+        recovery.newOwner = newOwner;
+        recovery.completeAfter = now + recoveryPeriod;
+        recovery.guardianCount = guardianCount;
+
+        securityStorage.setLock(wallet, now + lockPeriod);
+
+        emit RecoveryStarted(wallet, newOwner, recovery.completeAfter);
+    }
+
+    function cancelRecovery(
+        address   wallet,
+        address[] calldata signers
+        )
+        external
+        nonReentrant
+        onlyFromMetaTx
+    {
+        WalletRecovery storage recovery = wallets[wallet];
+        require(recovery.completeAfter > 0, "NOT_STARTED");
+
+        uint guardianCount = wallets[wallet].guardianCount;
+        require(signers.length >= (guardianCount + 1) / 2, "NOT_ENOUGH_SIGNER");
+        require(isWalletOwnerOrGuardian(wallet, signers), "UNAUTHORIZED");
+
+        delete wallets[wallet];
+        securityStorage.setLock(wallet, 0);
+
+        emit RecoveryCancelled(wallet);
+    }
+
+    function completeRecovery(address wallet)
+        external
+        nonReentrant
+        onlyFromMetaTx
```

Comment:
Does it need to be `onlyFromMetaTx`?

Reply:
The current design does not maintain the signature collection process in smart contract, the server/relayer will coordinate the collection of signatures, once enough signatures are collected, it will invoke the `startRecovery` method. But the transactor doesn't have to be a specific address, anyone including any guardian who has enough signatures can transact the meta-transaction.

Reply:
But `completeRecovery` doesn't need any signatures, no? I was just curious because it seems like anyone can complete the recovery now, so it seems like it doesn't matter if `completeRecovery` is called with a MetaTx or a direct call.

Reply:
You are right, I should remove onlyFromMetaTx so this method can be called by anybody.